File: backend/app/models/note.py
import uuid
from datetime import datetime, date
from sqlalchemy import String, Text, DateTime, Date, Float, ForeignKey, ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID
from pgvector.sqlalchemy import Vector
from app.db.base import Base


class VoiceNote(Base):
    __tablename__ = "voice_notes"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    audio_url: Mapped[str] = mapped_column(String(512), nullable=False)
    transcript: Mapped[str | None] = mapped_column(Text, nullable=True)
    title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    tags: Mapped[list[str] | None] = mapped_column(ARRAY(String), nullable=True)
    duration_seconds: Mapped[float | None] = mapped_column(Float, nullable=True)
    # "pending" | "processing" | "done" | "failed"
    transcription_status: Mapped[str] = mapped_column(
        String(20), default="pending", nullable=False
    )
    embedding: Mapped[list[float] | None] = mapped_column(Vector(1536), nullable=True)
    day_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    recorded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=datetime.utcnow
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    user: Mapped["User"] = relationship("User", back_populates="notes")

File: mobile/src/types/note.ts
export interface Note {
  id: string;
  user_id: string;
  audio_url: string;
  transcript: string | null;
  title: string | null;
  summary: string | null;
  tags: string[] | null;
  duration_seconds: number | null;
  transcription_status: "pending" | "processing" | "done" | "failed";
  day_date: string;
  recorded_at: string;
  created_at: string;
}

export interface NoteListItem {
  id: string;
  title: string | null;
  summary: string | null;
  tags: string[] | null;
  duration_seconds: number | null;
  transcription_status: string;
  day_date: string;
  recorded_at: string;
  audio_url: string;
}

export interface SearchResult {
  id: string;
  title: string | null;
  day_date: string;
  recorded_at: string;
  snippet: string;
  score: number;
}

export interface SourceNote {
  note_id: string;
  day_date: string;
  snippet: string;
}

export interface AskResponse {
  answer: string;
  sources: SourceNote[];
}

export interface ChatMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  sources?: SourceNote[];
  created_at: string;
}
